File: MaxProfit.py
#https://oj.leetcode.com/problems/best-time-to-buy-and-sell-stock/
#timing out on site but I think straight convert from the .java, logic should be ok 

class Solution:
    # @param prices, a list of integer
    # @return an integer
    def maxProfit(self, prices):
        
        length = len(prices)
        if length == 0 or length == 1:
        	return 0

        start = 0
        while start < length - 1 and prices[start] > prices[start + 1]:
        	start += 1

        if start == length - 1:
        	return 0

        profit = 0
        lower_price_exists = False
        keep_iterating = True

        while keep_iterating :
        	lower_price_exists = false

        	index = start + 1
        	while index in range(length):

        		if prices[start] < prices[index] and prices[index] - prices[start] > profit:
        			profit = prices[index] - prices[start]
        		elif prices[index] < prices[start] and not lower_price_exists:
        			start = index
        			lower_price_exists = True

        		index += 1

        	if not lower_price_exists or start == length - 2:
        		keep_iterating = False


        return profit

File: MaxProfit.java
//https://oj.leetcode.com/problems/best-time-to-buy-and-sell-stock/

public class Solution {
    public int maxProfit(int[] prices) {
        
        int length = prices.length;
        if(length == 0 || length == 1)
            return 0;
            
        int start = 0;
        while((start < length - 1) && (prices[start] > prices[start + 1])){
            start++;
        }

        if(start == length - 2)
            return 0;
        
        int profit = 0;
        boolean lowerPriceExists = false;
        boolean keepIterating = true;

        while(keepIterating){
            lowerPriceExists = false;

	        for(int i = start + 1; i < length; i++){
	        	if(prices[start] < prices[i] && (prices[i] - prices[start]) > profit)
	                profit = prices[i] - prices[start];
	            else if(prices[i] < prices[start] && !lowerPriceExists){
	            	start = i;
	            	lowerPriceExists = true;
	            }
	        }

	        if(!lowerPriceExists || start == length - 2)
	        	keepIterating = false;
	    }
	    return profit;
    }
}